[ BUG ] nil target item cause prometheus exit

**What did you do?**
using a file sd to define the target groups.

**What did you expect to see?**
wrong discovery data should be skipped or warning, not panic.

**What did you see instead? Under which circumstances?**
panic, and the details is:
```
INFO[0000] Starting target manager...                    source="targetmanager.go:63"

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x166053a]

goroutine 52 [running]:
github.com/prometheus/prometheus/discovery/file.readFile(0xc42015e2a0, 0x1b, 0x1, 0x1, 0xc420b4a270, 0x3, 0x3)
	/Users/Jacky/Jacky.Wu/MyProjects/Go/src/github.com/prometheus/prometheus/discovery/file/file.go:259 +0x1ea
github.com/prometheus/prometheus/discovery/file.(*Discovery).refresh(0xc420b28080, 0x3ea2080, 0xc420b28100, 0xc42001d620)
	/Users/Jacky/Jacky.Wu/MyProjects/Go/src/github.com/prometheus/prometheus/discovery/file/file.go:196 +0xfe
github.com/prometheus/prometheus/discovery/file.(*Discovery).Run(0xc420b28080, 0x3ea2080, 0xc420b28100, 0xc42001d620)
	/Users/Jacky/Jacky.Wu/MyProjects/Go/src/github.com/prometheus/prometheus/discovery/file/file.go:121 +0x198
created by github.com/prometheus/prometheus/discovery.(*TargetSet).updateProviders
	/Users/Jacky/Jacky.Wu/MyProjects/Go/src/github.com/prometheus/prometheus/discovery/discovery.go:249 +0x283
```

**Environment**

* System information:

	Linux 3.16.0-4-amd64 x86_64

* Prometheus version: 

```
 prometheus, version 1.6.0 (branch: master, revision: 10f645382988add09c8e3420d0d7ff2dc4d36a29)
  build user:       root@0816a56aa81c
  build date:       20170414-18:36:18
  go version:       go1.8.1
```

* Alertmanager version:

* Prometheus configuration file:
```
...
  - job_name: 'nil-item-array-json-job'
    metrics_path: /metrics
    file_sd_configs:
      - files:
         - "invalid_nil_item_array.json"
```

```
# invalid_nil_item_array.json
# !!! the last null line caused panic!

[
    {
        "labels": {
            "env": "prod"
        },
        "targets": [
            "localhost:9090"
        ]
    },
    null
]
```

--- a/discovery/file/file.go
+++ b/discovery/file/file.go
@@ -15,6 +15,7 @@ package file
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"path/filepath"
@@ -256,6 +257,11 @@ func readFile(filename string) ([]*config.TargetGroup, error) {
 	}
 
 	for i, tg := range targetGroups {
+		if tg == nil {
+			err = errors.New("nil target group item found")
+			return nil, err
+		}
+
 		tg.Source = fileSource(filename, i)
 		if tg.Labels == nil {
 			tg.Labels = model.LabelSet{}

--- a/discovery/file/file_test.go
+++ b/discovery/file/file_test.go
@@ -29,13 +29,17 @@ import (
 )
 
 func TestFileSD(t *testing.T) {
-	defer os.Remove("fixtures/_test.yml")
-	defer os.Remove("fixtures/_test.json")
-	testFileSD(t, ".yml")
-	testFileSD(t, ".json")
+	defer os.Remove("fixtures/_test_valid.yml")
+	defer os.Remove("fixtures/_test_valid.json")
+	defer os.Remove("fixtures/_test_invalid_nil.json")
+	defer os.Remove("fixtures/_test_invalid_nil.yml")
+	testFileSD(t, "valid", ".yml", true)
+	testFileSD(t, "valid", ".json", true)
+	testFileSD(t, "invalid_nil", ".json", false)
+	testFileSD(t, "invalid_nil", ".yml", false)
 }
 
-func testFileSD(t *testing.T, ext string) {
+func testFileSD(t *testing.T, prefix, ext string, expect bool) {
 	// As interval refreshing is more of a fallback, we only want to test
 	// whether file watches work as expected.
 	var conf config.FileSDConfig
@@ -56,13 +60,13 @@ func testFileSD(t *testing.T, ext string) {
 		t.Fatalf("Unexpected target groups in file discovery: %s", tgs)
 	}
 
-	newf, err := os.Create("fixtures/_test" + ext)
+	newf, err := os.Create("fixtures/_test_" + prefix + ext)
 	if err != nil {
 		t.Fatal(err)
 	}
 	defer newf.Close()
 
-	f, err := os.Open("fixtures/valid" + ext)
+	f, err := os.Open("fixtures/" + prefix + ext)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -80,8 +84,17 @@ retry:
 	for {
 		select {
 		case <-timeout:
-			t.Fatalf("Expected new target group but got none")
+			if expect {
+				t.Fatalf("Expected new target group but got none")
+			} else {
+				// invalid type fsd should always broken down.
+				break retry
+			}
 		case tgs := <-ch:
+			if !expect {
+				t.Fatalf("Unexpected target groups %s, we expected a failure here.", tgs)
+			}
+
 			if len(tgs) != 2 {
 				continue retry // Potentially a partial write, just retry.
 			}
@@ -90,12 +103,12 @@ retry:
 			if _, ok := tg.Labels["foo"]; !ok {
 				t.Fatalf("Label not parsed")
 			}
-			if tg.String() != filepath.FromSlash(fmt.Sprintf("fixtures/_test%s:0", ext)) {
+			if tg.String() != filepath.FromSlash(fmt.Sprintf("fixtures/_test_%s%s:0", prefix, ext)) {
 				t.Fatalf("Unexpected target group %s", tg)
 			}
 
 			tg = tgs[1]
-			if tg.String() != filepath.FromSlash(fmt.Sprintf("fixtures/_test%s:1", ext)) {
+			if tg.String() != filepath.FromSlash(fmt.Sprintf("fixtures/_test_%s%s:1", prefix, ext)) {
 				t.Fatalf("Unexpected target groups %s", tg)
 			}
 			break retry
@@ -135,7 +148,7 @@ retry:
 	}
 	newf.Close()
 
-	os.Rename(newf.Name(), "fixtures/_test"+ext)
+	os.Rename(newf.Name(), "fixtures/_test_"+prefix+ext)
 
 	cancel()
 	<-drained

--- a/discovery/file/fixtures/invalid_nil.json
+++ b/discovery/file/fixtures/invalid_nil.json
@@ -1,0 +1,9 @@
+[
+    {
+        "targets": ["localhost:9090", "example.org:443"],
+        "labels": {
+            "foo": "bar"
+        }
+    },
+    null
+]

--- a/discovery/file/fixtures/invalid_nil.yml
+++ b/discovery/file/fixtures/invalid_nil.yml
@@ -1,0 +1,5 @@
+- targets: ['localhost:9090', 'example.org:443']
+  labels:
+    foo: bar
+
+- null

